client/write_state.go

```diff
@@ -22,69 +22,24 @@ package client
 
 import (
 	"fmt"
-	"math"
 	"sync"
 
 	"github.com/m3db/m3cluster/shard"
-	"github.com/m3db/m3db/generated/thrift/rpc"
 	"github.com/m3db/m3db/topology"
 	xerrors "github.com/m3db/m3x/errors"
 	"github.com/m3db/m3x/ident"
 	"github.com/m3db/m3x/pool"
 )
 
-var (
-	writeOpZeroed = writeOp{shardID: math.MaxUint32}
-	// NB(bl): use an invalid shardID for the zerod op
-)
-
-type writeOp struct {
-	namespace    ident.ID
-	shardID      uint32
-	request      rpc.WriteBatchRawRequestElement
-	datapoint    rpc.Datapoint
-	completionFn completionFn
-}
+// writableOp represents a generic write operation
+type writableOp interface {
```

Comment:
As discussed, maybe to avoid the overloaded term we can go with `type writeOp interface` and `type writeOperation struct` to separate implementation vs interface.

Reply:
sure thing. done in next patch